### PR TITLE
Testorchestrator

### PR DESCRIPTION
**Pull request by KrzysiekKnigawka**
This is adding a way to run test described as json format list of test scenarios.
Tests scenario has flowing form:
{
    "FunctionName": "QueueTriggerNodeJS1",
    "Platform": "Azure",
    "InputPath": "TestSamples\\sqs\\input.txt",
    "TriggerType": "Queue",
    "LoadProfile": "Linear",
    "InputObject": "samples-node-js-cpu-perf-test",
    "OutputObject": "outqueue",
    "Eps": 60,
    "Repeat": true,
    "DurationInMinutes": 20,
    "WarmUpTimeInMinutes": 5
}

Tests are run sequentially in order described in the file.

Change is also adding an ILogger interface and two implementation of it:
Console (default)
File

File logger is used for running test scenarios, so results can be analyzed afterwards.

### CODE REVIEW

**Comment by paulbatum** [comment]
FYI I have an interviewing event today so I probably won't have a chance to look at this until next week. One nit: it looks really weird that the properties in your sample JSON are PascalCased. I'd expect them to be camelCased.

File: SampleUsages/ConsoleLogger.cs [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+
+        public void LogWarning(string format, params object[] args)
+        {
+            LogWarning(string.Format(format, args));
+        }
+
+        public void LogException(Exception ex)
+        {
+            var fg = Console.ForegroundColor;
+            Console.ForegroundColor = ConsoleColor.Yellow;
```
Comment (hforeste) [inline SampleUsages/ConsoleLogger.cs]: nitpick: Should this be foreground color of red?

Comment (KrzysiekKnigawka) [inline SampleUsages/ConsoleLogger.cs]: Yea, makes sense

File: SampleUsages/Program.cs [hunk 2]
```diff
@@ -19,43 +16,122 @@ public class Program : CmdHelper
...
+            using (StreamReader r = new StreamReader(scenarioFilePath))
+            {
+                string json = r.ReadToEnd();
+                try
+                {
+                    testScenarios = JsonConvert.DeserializeObject<List<TestScenario>>(json);
+                }
+                catch (JsonException ex)
+                {
+                    Console.WriteLine($"Failed to parse input file. {ex}");
```
Comment (hforeste) [inline SampleUsages/Program.cs]: either exit the execution here or throw an unhandled exception with the "Failed to parse input file" message

File: SampleUsages/TestScenario.cs [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+        internal void RunScenario(FunctionTest functionTest, TriggerTestLoadProfile profile)
+        {
+            var perfResult = functionTest.RunAsync(profile).GetAwaiter().GetResult();
+            _logger.LogInfo(perfResult.ToString());
+        }
+        
+        private void AssertInput(params string[] properties)
+        {
+            var fault = false;
+            var expcetions = new List<Exception>();
```
Comment (hforeste) [inline SampleUsages/TestScenario.cs]: nitpick: misspelling "exceptions"

File: ServerlessBenchmark/ServerlessPlatformControllers/AWS/AWS_Controller.cs [hunk 4]
```diff
@@ -16,6 +16,13 @@ namespace ServerlessBenchmark.ServerlessPlatformControllers.AWS
 {
     public class AwsController: ICloudPlatformController
     {
+        private ILogger _logger;
+
+        public AwsController(ILogger logger)
+        {
+            this._logger = logger;
+        }
```
Comment (hforeste) [inline ServerlessBenchmark/ServerlessPlatformControllers/AWS/AWS_Controller.cs]: Also do this for AzureController class. 
Maybe you can abstract the constructor away in a base class and initialize the logger there.

File: ServerlessBenchmark/ServerlessPlatformControllers/AWS/AWS_Controller.cs [hunk 5]
```diff
@@ -55,7 +62,7 @@ public CloudPlatformResponse PostMessages(CloudPlatformRequest request)
             var doneTask = await Task.WhenAny(aggregateTasks);
             if (doneTask.Id == timeoutTask.Id)
             {
-                Console.WriteLine("--ERROR-- Reached timeout {0}ms", timeout);
+                this._logger.LogWarning("--ERROR-- Reached timeout {0}ms", timeout);
```
Comment (hforeste) [inline ServerlessBenchmark/ServerlessPlatformControllers/AWS/AWS_Controller.cs]: >this._logger.LogWarning [](start = 16, length = 23)

nitpick: LogException

File: SampleUsages/TestScenario.cs [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+                    this._logger.LogException(ex);
+                    throw ex;
+            }
+
+            return profile;
+        }
+
+        internal void RunScenario(FunctionTest functionTest, TriggerTestLoadProfile profile)
+        {
+            var perfResult = functionTest.RunAsync(profile).GetAwaiter().GetResult();
```
Comment (KrzysiekKnigawka) [inline SampleUsages/TestScenario.cs]: I don't quite get how ConfigureAwait(continueOnCapturedContext: false) would help here. Accessing Result directly will swallow exceptions. A preferable way to use async method synchronously is to do it that way or Task.WaitAll() and handle everything manually.
It is wrapped with try{} catch(){} before it's used.

**Pull request merged**